### PR TITLE
Non-native field gadgets corrected

### PR DESCRIPTION
**Pull request by UlrichHaboeck75**
To summarize the content of this PR:

- **Documentation:** extensive documentation have been added inline. It explains in detail the rationale of each non-native arithmetic function, as well as its security bounds. As a by-product, we slightly optimized the choice of shift constants applied in some of the algorithms to prevent underflows of limb differences. 
- **Clarified normalization strategy:** We switched the automatic normalization strategy from the initial post-reduction approach to an even more greedy pre-reduction paradigm (with multiplication results as the only exception).  We clarify the design by separating the arithmetic functions into 
  - (private) low-level functions, which do not make use of any reduction, and
  - (public) high-level functions, which perform a pre-reduction step on the operands (if needed), before calling the corresponding low-level function.
- **Patching security issues**: We corrected the choice of the limb size bounds (the `num_of_additions_over_normal_form`) in some of the arithmetic functions. 
- **Thorough tests**: Previous tests did not provocate edge cases on oversized limbs in a controlled manner.  We introduce a separate allocation gadget for non-native gadgets with oversized limbs and use it in the elementary tests of the arithmetic functions. Furthermore, we clarified the general test design by a bottom-up approach, from tests for low-level functions (equality checks and modular reduction), to elementary test of modular arithmetics,  and eventually the stress tests.

### CODE REVIEW

**Review by 95DDB — APPROVED** [review]

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 1]
```diff
@@ -396,22 +397,23 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
-        // TODO: panic if 
-        // ``
-        //      2 * bits_per_limb + surfeit(product) + len(num_limbs) > CAPACITY - 2.
+        //     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
         // ``
+        // with `surfeit'` as above.
 
         let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
-        let num_add_bound = ConstraintF::from(params.num_limbs as u64) 
+        let num_add_bound = ConstraintF::from((params.num_limbs as u64)^2)
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: This is not a square operation. You are doing a XOR here

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 2]
```diff
@@ -497,21 +499,20 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
-        //      2 * bits_per_limb + surfeit(product) + len(num_limbs) > CAPACITY - 2.
+        //     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
         // ``
+        // with `surfeit' = len(num_limbs^2 * (num_add(L)+1) + 1)`.
 
         let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
-        let num_add_bound = ConstraintF::from(params.num_limbs as u64) 
-            * (self.num_of_additions_over_normal_form + ConstraintF::one());
-        let surfeit_product = bitlen!(num_add_bound);
+        let num_add_bound = ConstraintF::from((params.num_limbs as u64)^2)
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Again, this is not a square it's a XOR

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 3]
```diff
@@ -279,17 +280,20 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        // Hence the surfeit for the limbs of `k*p + r` is 
+        // ``
+        //      surfeit(kp + r) <= len(num_limbs * (num_adds + 1) + 1)
         // ``
         let mut kp_plus_r_gadget = Self {
             limbs: prod_limbs,
             num_add_over_normal_form: ConstraintF::from(params.num_limbs as u64) 
-                * (p_gadget.num_of_additions_over_normal_form + ConstraintF::one())
-                * (k_gadget.num_of_additions_over_normal_form + ConstraintF::one()) 
-                + ConstraintF::one(),
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: This last +1 should've been kept ?

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 4]
```diff
@@ -217,19 +217,24 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // ``
+        //      surfeit' = len(num_limbs^2 * (num_add(L)+1) * (num_add(R) + 1) + 1).
         // ``
         Self::reduce_until_cond_is_satisfied(
             cs.ns(|| "pre mul reduce"),
             elem,
             elem_other,
             |elem, elem_other| {
-                let num_add_bound = ConstraintF::from(params.num_limbs as u64) 
+                let num_add_bound = ConstraintF::from((params.num_limbs as u64)^2)
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: This is not a square, it's a XOR

**Review by 95DDB — CHANGES_REQUESTED** [review]

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 1 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: done

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 2 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: done

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 4 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: done

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 3 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: Well, the formula was not correct anyway.

**Review by 95DDB — APPROVED** [review]

**Comment by UlrichHaboeck75** [comment]
To summarize the content of this PR:

- **Documentation:** extensive documentation have been added inline. It explains in detail the rationale of each non-native arithmetic function, as well as its security bounds. As a by-product, we slightly optimized the choice of shift constants applied in some of the algorithms to prevent underflows of limb differences. 
- **Clarified normalization strategy:** We switched the automatic normalization strategy from the initial post-reduction approach to an even more greedy pre-reduction paradigm (with multiplication results as the only exception).  We clarify the design by separating the arithmetic functions into 
  - (private) low-level functions, which do not make use of any reduction, and
  - (public) high-level functions, which perform a pre-reduction step on the operands (if needed), before calling the corresponding low-level function.
- **Patching security issues**: We corrected the choice of the limb size bounds (the `num_of_additions_over_normal_form`) in some of the arithmetic functions. 
- **Thorough tests**: Previous tests did not provocate edge cases on oversized limbs in a controlled manner.  We introduce a separate allocation gadget for non-native gadgets with oversized limbs and use it in the elementary tests of the arithmetic functions. Furthermore, we clarified the general test design by a bottom-up approach, from tests for low-level functions (equality checks and modular reduction), to elementary test of modular arithmetics,  and eventually the stress tests.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 5]
```diff
@@ -1402,150 +1811,178 @@ macro_rules! nonnative_test_individual {
 
 macro_rules! nonnative_test {
     ($test_name:ident, $test_simulation_field:ty, $test_constraint_field:ty) => {
+        // Commented out as the estimated number of constraints from `find_paramaters`
+        // slightly differ from the measured ones.
+        //
+        // elementary_test!(
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs:1817]: Can we remove it then ?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 6]
```diff
@@ -1602,32 +2039,62 @@ macro_rules! nonnative_test {
...
+#[cfg(all(feature = "tweedle", feature = "ed25519"))]
+nonnative_test!(ed25519Fq_over_TweedleFr, ed25519Fq, TweedleFr);
+
+#[cfg(all(feature = "tweedle", feature = "secp256k1"))]
+nonnative_test!(secp256k1_over_TweedleFr, secp256k1Fq, TweedleFr);
+
+#[cfg(all(feature = "tweedle", feature = "bn_382"))]
+nonnative_test!(Bn382Fr_over_TweedleFr, Bn382Fr, TweedleFr);
+
+#[cfg(all(feature = "tweedle", feature = "mnt4_753"))]
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs:2059]: Are these mnt4 tests necessary ? They are quite heavy and we will remove the curve anyway

**Review by 95DDB — CHANGES_REQUESTED** [review]

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 7]
```diff
@@ -179,211 +569,445 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        // where 
+        // ``
+        //      surfeit(product) = log(num_limbs * (num_add(L)+1) * (num_add(R) + 1)).
+        // ``
+        // To allow for a subsequent reduction we need to assure the stricter condition 
+        // that 
+        // ``
+        //     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+        // ``
+        // with `surfeit' = log(num_limbs * (num_add(L) + 1) * (num_add(R) + 1) + num_limbs)`.
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs:635]: Is surfeit' = surfeit(k*p + r)? In that case surfeit' = len( num_limbs + 2^(len(num_adds +1)) )
or = len( num_limbs + num_adds + 1 ) according to Ulrich's estimate.
surfeit_prod = len( (num_add(L) + 1) * (num_add(R) + 1) * num_limbs ) is the extra amount of extra bits that
the limbs of the product have respect to 2 * bits_per_limb.
I think that the final surfeit is underestimated now. In fact, in order to do the reduction,
we need that length of a limb of (L * R) + (k * p + r) + carry  does not exceed CAPACITY. Recall that
k has all the limbs in normal form but the most significant one which requires bits_per_limb + surfeit_prod bits.
An upper bound for (L * R) + (k * p + r) is
num_limbs * (2 ^ (2 * bits_per_limb + surfeit_product) - 1 ) ^ 2 + ( num_limbs - 1) * (2^bits_per_limb - 1)^2 + (2^(bits_per_limb + surfeit_prod) - 1 ) * (2^bits_per_limb - 1).  It follows that we can estimate  the "final" 
surfeit = surfeit_prod + 1.
    surfeit = len(num_limbs * (num_add(L)+1) * (num_add(R) + 1)) + 1.
len( num_limbs * (num_add(L) + 1) * (num_add(R) + 1) + (num_limbs - 1) )
DOUBT: It is a bit confusing that the reduce function for NonNativeFieldMulResult has as argument only the surfeit of k*p + r
and it does not depend on the surfeit of the product L * R.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 8]
```diff
@@ -179,211 +569,445 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+    /// This intermediate representations can be added with each other, and they can later be
+    /// reduced back to the `NonNativeFieldGadget`.
+    /// Assumes that
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+    /// ``
+    /// where
+    /// ``
+    ///      surfeit' = log(num_limbs^2 * (num_add(L)+1) + 1).
+    /// ``
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs:725]: Specify in the inline comment: The constant is reduced in normal form: this is why it does not appear num_add(R)..

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 9]
```diff
@@ -76,13 +131,69 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(res)
     }
 
-    /// Constraints for reducing the result of a multiplication mod p, to get an original representation.
+    /// Reducing a NonNativeMulResultGadget back to a non-native field gadget
+    /// in normal form. Assumes that 
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2.
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:137]: If really, as described in group_and_check_equality, the carries don't affect length bounds, we should have <= CAPACITY - 1. Anyway now I think that carries **do** affect the bounds, so it is fine.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 10]
```diff
@@ -76,13 +131,69 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+    /// ``
+    /// where 
+    /// ``
+    ///     bits_per_limb = NonNativeFieldParams::bits_per_limb, 
+    ///     surfeit' =  log(num_limbs + num_adds + 1), 
+    ///     num_limbs = NonNativeFieldParams::num_limbs,
+    /// ``
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:144]: Better to call surfeit' = surfeit(kp + r), it is more clear.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 11]
```diff
@@ -76,13 +131,69 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+    /// and `num_adds` is as in the NonNativeMulResultGadget.
+    // Costs
+    // ``
+    //     C =  len(p) + surfeit' 
+    //          + len(p) + num_limbs^2
+    //          +  (num_groups - 1) * (3 + bits_per_limb + surfeit') + 1
+    // ``
+    // constraints, with `num_groups = Ceil(len(p) / S)` and
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:152]: S is the number of limbs in a group, so num_groups should be 
num_groups =  ceil(bit_len(prod) / (S * bits_per_limb)). 
Note that max_bit_len(prod) = 2*len(p) + len(num_adds(prod) + 1).

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 12]
```diff
@@ -76,13 +131,69 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        //      len(k) <= len(p) + len(num_adds + 1). 
+        // `` 
         let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
 
+        // Note: To assure that the limb-wise computation of the `k * p + r` does not 
+        // exceed the capacity bound, we demand 
+        // ``
+        //     2 * bits_per_limb + surfeit'  <= CAPACITY,
+        // ``
+        // with `surfeit' =  log(num_limbs + num_adds + 1)`.
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:190]: This is not completely correct in my opinion: it should be len((num_limbs + 2^(len(num_adds + 1)) ). I will explain this more precisely below.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 13]
```diff
@@ -99,15 +210,15 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
             simulation_phantom: PhantomData,
         };
 
         // Step 2: compute surfeit
-        let surfeit = overhead!(self.prod_of_num_of_additions + ConstraintF::one()) + 1 + 1;
+        let surfeit = ceil_log_2!(BigUint::one() + &self.num_add_over_normal_form);
 
-        // Step 3: allocate k
+        // Step 3: allocate k,
+        // Costs `C = len(p) + surfeit` constraints.
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:221]: It costs len(p) + len(num_adds + 1) + 1, which is bigger than len(p) + len(num_adds + 1) as it was claimed in a comment above and it is practically used in the code:
        surfeit = ceil_log_2!(BigUint::one() + &self.num_add_over_normal_form)
WARNING: Recall that the reduced inputs are < 2^len(p) - 1, but they can be > p, that's why we need to add +1 in my opinion.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 14]
```diff
@@ -165,21 +281,24 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
             NonNativeFieldGadget::<SimulationF, ConstraintF>::alloc(cs.ns(|| "alloc r"), || {
                 self.value()
             })?;
 
-        let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
-
-        // Step 1: reduce `self` and `other` if necessary
+        // Compute the product representation for `k*p`
+        // Costs `C =  num_limbs^2` constraints.
+        // TODO: let us use `mul_by_constant()` and `add_by_constant()`. Costs no constraints.
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:301]: More precisely we can substitute that constraints with linear combinations.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 15]
```diff
@@ -205,13 +324,27 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        //      (k * p + r)[i] <= (num_limbs + num_add)* 2^{2*bits_per_limb[i]} 
+        //                      + 2^bits_per_limb[i] 
+        //                      <= (num_limbs + num_add + 1) * 2^{2*bits_per_limb[i]}. 
+        // ``
+        // Hence we may set num_adds and surfeit for the limbs of `k*p + r` according
+        // to
+        // ``
+        //      num_adds(kp + r) = num_limbs + num_add,
+        //      surfeit(kp + r) = log(num_limbs + num_adds + 1).
+        // ``
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:340]: This can be explained more precisely in my opinion:
the addends limb of   k * p   are at (num_limbs - 1) which are product of two limbs in normal form
 and at most one which is a product of one in normal form and one oversized.
(k * p + r)[i] <= 2^bits_per_limb -1   +   (num_limbs - 1) * (2^bits_per_limb - 1)^2 
                + (2^bits_per_limb - 1) * (2^(bits_per_limb + len(num_adds + 1) + 1) - 1)
                     <  (num_limbs - 1) * 2^(2 * bits_per_limb ) + 2^(2 * bits_per_limb + len(num_adds + 1) + 1)
                     =  2^(2 * bits_per_limb) * ( num_limbs - 1 + 2^( len(num_adds + 1) ) + 1)
From this it would follow that surfeit(k*p + r) = len(num_limbs + 2^( len(num_adds + 1) ).
WARNING: Note that this can be bigger than len(num_limbs + num_adds + 1).

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 16]
```diff
@@ -227,10 +360,30 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        //          (ConstraintF::CAPACITY - 2 - (2 * bits_per_limb + surfeit') / bits_per_limb
+        //      ] 
+        //        = Floor[
+        //          (ConstraintF::CAPACITY - 2 - surfeit') / bits_per_limb
+        //      ] - 2.
+        // ``
         Reducer::<SimulationF, ConstraintF>::group_and_check_equality(
             cs.ns(|| "group and check equality"),
-            surfeit,
+            surfeit_kp_plus_r,
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:383]: Why does this depend only on the surfeit of k * p + r?

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 17]
```diff
@@ -98,110 +76,294 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // ``
+        // To assure that the following operations on the totals do not exceed the capacity, 
+        // it is sufficient to demand 
+        // ``
+        //      bits_per_group + 1 <= ConstraintF::CAPACITY,
+        // ``
+        // as described in detail below. This yields the security condition
+        // ``
+        //      bits_per_limb + surfeit + (S - 1) * shift_per_limb + 2 <= ConstraintF::CAPACITY,
+        // ``
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs:336]: I would specify in the comment that the value below is the maximum S which achieves the inequality above.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 18]
```diff
@@ -252,25 +421,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // and hence
+        // ``
+        //      L[i] + (shift_constant - R[i]) < 2^{bits_per_group + 1},
+        // ``
+        // Since the length of the carries are throughout `<= bits_per_limb + surfeit`, 
+        // we also have `carry[i-1] + L[i] < 2^bits_per_group`, and the overall sum 
+        // ``
+        //      carry[i-1] + L[i] + (shift_constant - R[i]) 
+        // `` 
+        // is still of length `<= bits_per_group + 1`.
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs:449]: I am not sure about that. Below a more detailed explanation.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 19]
```diff
@@ -252,25 +421,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // `` 
+        //   carry + limb[0] + limb[1] * A < 2 * A * 2^limb_size
+        // ``
+        // and therefore 
+        // ``
+        //   carry + limb[0] + limb[1] * A + limb[2] * A^2 < 2 * A^2 * 2^limb_size  
+        //   ...
+        //   carry + limb[0] + limb[1] * A + ... + limb[S-1] * A^{S-1} <
+        //                       < 2 * A^{S-1} * 2^limb_size.
+        // ``
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs:462]: I believe this inequality, I do not understand the implication below.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 20]
```diff
@@ -252,25 +421,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // we also have `carry[i-1] + L[i] < 2^bits_per_group`, and the overall sum 
+        // ``
+        //      carry[i-1] + L[i] + (shift_constant - R[i]) 
+        // `` 
+        // is still of length `<= bits_per_group + 1`.
+        
+        // Why carries don't effect the length bound: 
+        // Assuming `len(carry) <= bits_per_limb + surfeit`, which is shown in detail below, 
+        // `` 
+        //   carry + limb[0] + limb[1] * A < 2 * A * 2^limb_size
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs:454]: I do not understand this inequality: it should be < carry + (1 + A) * 2^limb_size. I accept anyway that carry < 2^(limb_size - shift_per_limb + 1), so the inequality holds a fortiori.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 21]
```diff
@@ -252,25 +421,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        //   carry + limb[0] + limb[1] * A + limb[2] * A^2 < 2 * A^2 * 2^limb_size  
+        //   ...
+        //   carry + limb[0] + limb[1] * A + ... + limb[S-1] * A^{S-1} <
+        //                       < 2 * A^{S-1} * 2^limb_size.
+        // ``
+        // Thus
+        // ``
+        //   len(carry + L[i]) <= 
+        //              limb_size + 1 + (S-1) * len(A) = bits_per_group.
+        // ``
```
Comment (DDT92) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs:467]: This explanation does not convince me and I do not understand the calculations. In my opinion the carries can effect
length bounds: suppose for example that A = 2^2 and we have
L = (0)(111)(100), L_red = (10)(00)(00), where the second is already reduced in normal form.__rust_force_expr!
In this case carry[0] = 1, L[1] + carry[0] = 1000, thus we exceed the length bound.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 5 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs:1817]: I'd keep it for later. On the other hand we can always restore it from the current commit.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 6 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs:2059]: The more fields we test the better.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 15 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:340]: As discussed offline, the limbs of kp are strictly bounded by `num_adds + 1 + num_limbs -1` times `2^{2*bits_per_limb}`. The `+1` in the stated formula stems from the addition with `r`.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 16 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:383]: Because `num_adds + num_limbs + 1` is bigger than `num_adds`.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 13 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:221]: Why it should cost more than `len(p) + ceil_log2(num_adds+1)` constraints? The quotient `k` is strictly bounded by `num_adds +1 ` times `2^len(p)`, otherwise the product `kp` und thus `kp+r` is too large to be reached by a mul result having `num_adds`. Said this, `k` has at most `ceil_log2(num_adds+1) + len(p)` bits.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 12 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs:190]: I explained below why this is correct.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 18 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs:449]: The fact that the carry doesn't cause an additional increase of the length is not obvious. Let's discuss it offline if my comments on this do not help.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 20 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs:454]: As said above, let's discuss it offline.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 21 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs:467]: As above, let's discuss it offline.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 7 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs:635]: I explained it below. If that doesn't help, let's discuss it offline.
